docs/man/rpm-dependency-generators.7.scd

```diff
@@ -0,0 +1,381 @@
+macro expansion itself is considered as the generated dependencies, one per
+line. As with traditional generators, the path passed as *%1* has *%buildroot*
+prepended.
+
+While the RPM macro primitives are limited, using *%{lua:...}* enables the
+writing of full-featured, macro-only generators in the Lua language.
+
+The following trivial example creates provides from path basenames, which is
+enormously faster than the equivalent traditional generator shelling out to
+execute a script that calls *basename*(1):
+
+```
+%__foo_provides()	%{basename:%{1}}
+```
+
+Added: 4.16.0
+
+## Multifile protocol
+
+Generators may optionally support an enhanced mode where all matching files are
+passed to the generator at once, and the generator is expected to print the
+processed filename prepended with *;* before printing the dependencies of that
+file. Files with no dependencies may be omitted.
+
+This is several orders of magnitude faster than the traditional behavior where a
+generator is launched separately for each and every file with the matching file
+attribute. To enable multifile mode for a generator, define the following macro:
+
+```
+%__NAME_protocol multifile
+```
+
+Added: 4.20.0
+
+# PACKAGE-LEVEL DEFINITIONS AND TWEAKS
+
+Sometimes, it may be necessary to tweak an installed generator for a specific
+package, or to use one that's shipped by the package itself also during its own
+build process. This can be done directly in an *rpm-spec*(5) file or by running
+*rpmbuild*(1) with the *--define* option(s). The following subsections apply to
+both methods.
+
+## File attributes and generators
+
+To introduce new file attributes and generators for the duration of the build,
+define them as normal and register them with the following macro (_NAMES_ is a
+colon-separated list of file attribute names):
+
+```
+%_local_file_attrs NAMES
+```
+
+It may be handy to use an executable in the *%{buildroot}* or a source file with
+these generators. See *EXAMPLES* for a concrete example.
+
+Added: 4.20.0
+
+## Excludes
+
+Unwanted dependencies can be filtered out with the following set of macros:
+
+```
+%__requires_exclude
+%__requires_exclude_from
+%__provides_exclude
+%__provides_exclude_from
+```
+
+The values are macro-expanded on use and then interpreted as extended regular
+expressions.
+
+The *exclude_from* variants operate on paths. For example, the following macro
+will prevent all provides generated from _\*.so_ files in *%{\_libdir}/mypkg/*
+from being added:
+
+```
+%define __provides_exclude_from ^%{_libdir}/mypkg/.*.so$
+```
+
+The *exclude* variants operate on generated dependency strings. For example, the
+following macro will prevent all typical library requires from being added,
+regardless of which files they originate from:
+
+```
+%define __requires_exclude ^lib.*$
+```
+
+Note that, within spec files, any backslashes need to be double-escaped in order
+to prevent the spec parser from consuming them.
```

Comment:
As a special case there's also %__global_DEPTYPE_exclude_from, which is not something folks should be using from the generators or specs, but maybe it should be mentioned from the POV that such a thing exists because it can and does affect the generator output, and so can be a bit baffling in the "why isn't my generator working for X".